mono/mini/aot-mangler.c

```diff
@@ -0,0 +1,1244 @@
+/**
+ * \file
+ * Aot Mangler Functions
+ *
+ * Authors:
+ *   Alexander Kyte (alkyte@microsoft.com)
+ *
+ * Copyright 2017 Microsoft, Inc (http://www.microsoft.com)
+ * Licensed under the MIT license. See LICENSE file in the project root for full license information.
+ */
+
+#include <mono/utils/json.h>
+#include "config.h"
+#include "aot-compiler.h"
+
+/*
+ * Mono Method Mangler Format:
+ *
+ * It is often useful to have a unique string associated with
+ * each method. The complications with using this to describe so
+ * rich a type system as C# is the issue of transitive type references.
+ *
+ * It is necessary to bake in information about the referenced types which 
+ * becomes quite verbose when they nest. Clarity around the depth of the tree
+ * at a given point in the string is important. We elect to use escaped brackets.
+ *
+ * Without using special characters, this is pretty difficult to do without
+ * risking a clash with user-provided names. For this reason we encode the mangled
+ * name in a format that is later scrubbed of these characters. This necessitates that we
+ * make our emitted replacements unique, unable to clash with user-provided names. In our case
+ * this is done by escaping the separator character we insert, the underscore.
+ *
+ * This makes for long, difficult-to-read strings with lots of structural nesting.
+ * 
+ * This motivated the creation of a demangler that prints verbose information about the
+ * mangled string in question.
+ *
+ *
+ * example mangled string:
+ *     mono_underscore_aot_space__lbrack__space_method_space_asm:_space_mscorlib_space__lbrack__space_klass_space_namespace:_space_System_space_name:_space_byte_space__rbrack__space_System_dot_IConvertible_dot_ToUInt32_space__lbrack__space_sig_space__lbrack__space_type_space_u4_space__rbrack__space_hasthis_space__lbrack__space_type_space_other_space_System_dot_IFormatProvider_space__rbrack__space__rbrack__space__rbrack__space_
+ * 
+ *
+ * example demangler output:
```

Comment:
This json dump is quite verbose. It's more easily parseable, but not necessarily easier to understand.
Maybe I'm missing the point, is there an use for it beyond us reading it?

So, I manually desanitized the example. It results in:
```
mono_aot [ method asm: mscorlib [ klass namespace: System name: byte ] System.IConversible.ToUInit [ sig [ type u4 ] hasthis [ type other System.IformatProvider ] ] ]
```
The above already has reasonable clarity and is denser than json.

One further improvement would be to make it closer to conversional method stringification.
```
method:  [mscorlib] System.byte::System.IConversible.ToUnit(uint, System.IFormatProvider)
````

Demangling is to help us understand what's going on.

TBH, I'm not sure the second conversion is that important.